support for ttn.v3 added, TTN node geo data stored in json

Peter-dM opened the pull request.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -138,16 +152,18 @@ function node_button($id) {
 
-		$filter_url_html = htmlspecialchars('?gateways=' . urlencode($gw_id) . '&show_other_gateways=1');
+		#$filter_url_html = htmlspecialchars('?gateways=' . urlencode($gw_id) . '&show_other_gateways=1');
+		$filter_url_html = htmlspecialchars('?gateways=' . urlencode($gw_id) . '&show_other_gateways=0');
```
Heb je show_other_gateways hier bewust op 0 gezet, of was dat alleen om te testen?

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -166,33 +182,68 @@ function gateway_button($gw_id, $gw_data = null) {
-					$meta['gtw_id'] = 'eui-' . strtolower($meta['gateway_eui']);
-					$meta['snr'] = $meta['lsnr'];
-					$gateways[] = $meta;
+				// fix old v2 record-layout:
+				if (!array_key_exists('gateways', $metadata)) {
+					$gateways = [];
+	
+					// Convert TTN staging to production format
+					foreach ($metadata as $meta) {
+						$meta['gtw_id'] = 'eui-' . strtolower($meta['gateway_eui']);
+						$meta['snr'] = $meta['lsnr'];
+						$gateways[] = $meta;
+					}
+	
+					$metadata = [
+						'gateways' => $gateways,
+						'frequency' => $metadata[0]['frequency'],
+						'data_rate' => $metadata[0]['datarate'],
+						'coding_rate' => $metadata[0]['codingrate'],
+					];
```
Dit blokje code converteert van v1 naar v2, maar staat nu alleen binnen het source == ttn.v2 blok, dus dat klopt nog niet. Volgens mij moet het iets worden als:

```
if (ttn.v1) {
  // Bovenstaand blok
} else if (ttn.v2) {
  $metadata = $message['metadata'];
} else if (ttn.v3) {
 // Jouw nieuwe code
}
```

matthijskooijman replied:
Of misschien is het nog explicieter om in het geval van ttn.v2 het uit te schrijven:
```
			$metadata = [
					'gateways' => $message['metadata']['gateways'],
					'frequency' => $message['metadata']['frequency'],
					'data_rate' => $message['metadata']['datarate'],
					'coding_rate' => $message['metadata']['codingrate'],
			];
```

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -166,33 +182,68 @@ function gateway_button($gw_id, $gw_data = null) {
+				$metadata = $message['uplink_message']['rx_metadata'];
+				$settings = $message['uplink_message']['settings'];
+	
+				if (!array_key_exists('gateway_ids', $metadata)) {
```
Volgens mij is deze if bij ttn.v3 niet meer relevant, toch? Ik denk dat die key nooit bestaat (dus if altijd true), dus die kan gewoon weg?

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -166,33 +182,68 @@ function gateway_button($gw_id, $gw_data = null) {
+				if (!array_key_exists('gateway_ids', $metadata)) {
+					$gateways = [];
+	
+					// Convert TTN staging to production format
```
Dit commentaar is niet correct, staging was ttn.v1, production was ttn.v2, dit converteert ttn.v3 naar ttn.v2 formaat voor verder gebruik in de rest van het script.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -166,33 +182,68 @@ function gateway_button($gw_id, $gw_data = null) {
+	
+					// Convert TTN staging to production format
+					foreach ($metadata as $meta) {
+						$meta['gtw_id'] = strtolower($meta['gateway_ids']['gateway_id']);
```
Is deze strtolower nodig? Ik zou denk ik liever het id onveranderd laten. In de ttn.v1 -> ttn.v2 conversie gebeurt het omdat de conventie van v2 is EUI-based gateways een id van een lowercase eui te geven, maar dat is hier niet van toepassing. Ik zou deze `strtolower` dus weglaten.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -166,33 +182,68 @@ function gateway_button($gw_id, $gw_data = null) {
+							$meta['gtw_eui'] = strtolower($meta['gateway_ids']['eui']);
+						if ($meta['gtw_id'] == 'packetbroker')
+							$meta['gtw_id'] = 'unknown';
+						if (! isset($meta['snr'])) 
```
Hier staat een spatie achter de ! teveel.

matthijskooijman replied:
Idem voor een aantal andere plekken elders in het script.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -222,6 +273,51 @@ function output_cell($rowspan, $data) {
+		// enhance gwdata with alias and geo
+		$gateways2 = array();
+		foreach ($gateways as $gwdata) {
+			if ($deb > 1) {
+				print("310 ");
+				print_r($gwdata);
+				print("<br>");
+			}
```
Dit is debug code, die mag er uit voor we dit mergen.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -222,6 +273,51 @@ function output_cell($rowspan, $data) {
+				print("<br>");
+			}
+			// add alias if missing
+			if (! isset($gwdata['alias']))
+				$gtw_id = $gwdata['gtw_id'];
```
Kan `$gwdata['alias']` hier uberhaupt geset zijn? De code hierboven zet nergens een alias, toch? Dus volgens mij kan de if weg, en de `$gtw_id = ...` regel blijven.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -222,6 +273,51 @@ function output_cell($rowspan, $data) {
+			// add alias if missing
+			if (! isset($gwdata['alias']))
+				$gtw_id = $gwdata['gtw_id'];
+				if (isset($ttn_gw_data[$gtw_id]['alias'])) 
+					$gwdata['alias'] = $ttn_gw_data[$gtw_id]['alias'];
```
Deze regels hebben de verkeerde indentation (of er mist een `{` en `}` als deze if binnen de if erboven hoort.

Daarnaast: zijn deze twee regels uberhaupt nodig? Volgens mij staan ze hieronder (regel 293/294) nog een keer identiek, en zijn ze (binnen de `if(!unknown)` beter op hun plek.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -222,6 +273,51 @@ function output_cell($rowspan, $data) {
+			if ($gwdata['gtw_id'] != 'unknown') {
+				if (! isset($gwdata['longitude']) || ! isset($gwdata['latitude'])) {
+					$gtw_id = $gwdata['gtw_id'];
+					if (isset($ttn_gw_data[$gtw_id]['alias'])) 
+						$gwdata['alias'] = $ttn_gw_data[$gtw_id]['alias'];
```
Ik zou deze twee regels een `if` naar buiten verplaatsen, zodat ze ook gebeuren als er wel een longitude en latitude zijn.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -222,6 +273,51 @@ function output_cell($rowspan, $data) {
+			// add geo if missing
+			if ($gwdata['gtw_id'] != 'unknown') {
+				if (! isset($gwdata['longitude']) || ! isset($gwdata['latitude'])) {
+					$gtw_id = $gwdata['gtw_id'];
```
Dit gebeurt ook al hierboven (regel 286), dus volgens kan deze regel weg.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -222,6 +273,51 @@ function output_cell($rowspan, $data) {
+
+			$gateways2[] = $gwdata;
+
+			if ($deb > 0 && $gwdata['latitude'] != 0) {
+				print("350 ");
+				print_r($gwdata);
+				print("<br>");
+			}
```
Ook weer debug code.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -222,6 +273,51 @@ function output_cell($rowspan, $data) {
 
+		// enhance gwdata with alias and geo
+		$gateways2 = array();
+		foreach ($gateways as $gwdata) {
```
Je zou hier een foreach by reference kunnen gebruiken, dan kun je direct de originele `$gateways` aanpassen door dat `$gwdata` dan een referentie wordt naar de data in  `$gateways`, je hebt `$gateways2` dan niet nodig. Volgens mij is de syntax daarvoor:

```PHP
foreach ($gateways as &$gwdata) {
```

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -278,11 +381,16 @@ function output_cell($rowspan, $data) {
 					output_cell($rowspan, "<a href=\"" . $url . "\">" . $row["latitude"] . " / " . $row["longitude"] . "</a>");
 				}
+
 				if (array_key_exists('counter', $message)) {
 					output_cell($rowspan, $message['counter']);
+				} else if (array_key_exists('f_cnt', $message['uplink_message'])) {
+					output_cell($rowspan, $message['uplink_message']['f_cnt']);
```
Hier ben je eigenlijk afhankelijk van ttn.v1/v2 vs ttn.v3 een andere element uit `$message` aan het pakken voor de frame counter. Het zou denk ik netter zijn om alle ttn.v1/v2/v3 code naar bovenaan te verplaatsen en daar `$metadata['counter']` oid te vullen, zodat je hier gewoon direct alleen die waarde hoeft te gebruiken. Volgens mij betekent dat ook dat je alleen in de v1/v2/v3 code bovenaan iets met `$message` hoeft te doen, en je daarna alleen nog maar naar `$row`, `$metadata` en `$gateways` hoeft te kijken.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -292,18 +400,36 @@ function output_cell($rowspan, $data) {
-				echo("  <td>" . gateway_button($gw_id, $gwdata) . "</td>\n");
+				if ($gw_id == "unknown") {
+					echo("  <td><I>" . $gw_id . "</i></td>\n");
+				} else if (($gwdata['longitude'] == 0) && ($gwdata['latitude'] == 0)) {
+					echo("  <td>" . $gw_id . "</td>\n");
```
Je roept hier nu `gateway_button` niet aan als er geen locatie is voor de gateway, maar dat betekent dat je dan ook niet op gateway id kunt filteren. Daarnaast checkt `gateway_button` zelf al of er een locatie is of niet, dus volgens mij kunnen deze twee regels weg.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -292,18 +400,36 @@ function output_cell($rowspan, $data) {
+				if ($row['source'] == 'ttn.v2')
+					echo("  <td>" . htmlspecialchars($metadata["frequency"]) . "MHz, " . htmlspecialchars($metadata["data_rate"]) . ", " .htmlspecialchars($metadata["coding_rate"]) . "CR</td>\n");
+				if ($row['source'] == 'ttn.v3')
+					echo("  <td>" . htmlspecialchars($metadata["frequency"]) . "MHz, SF" . htmlspecialchars($metadata["spreading_factor"]) . "BW" . htmlspecialchars($metadata["data_rate"]) . ", " .htmlspecialchars($metadata["coding_rate"]) . "CR</td>\n");
```
Ik zou deze conversie (`SF`, `spreading_factor`, `BW` en `data_rate` aan elkaar plakken) bovenaan doen, waar je toch al v3 apart behandeld en het resultaat in `$metadata["data_rate"]` stoppen. Je kunt dan hier gewoon onafhankelijk van de source versie gewoon `$metadata["data_rate"]` direct gebruiken (dus zoals het er origineel stond).

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -357,11 +484,18 @@ function output_cell($rowspan, $data) {
+				$alias = "&nbsp;";
+				if (isset($ttn_gw_data[$gw_id]['alias']))
+					$alias = $ttn_gw_data[$gw_id]['alias'];
+				$url = "https://ttnmapper.org/gateways/?gateway=$gw_id&startdate=&enddate=&gateways=on&points=on";
```
Goed idee om te linken naar ttnmapper, misschien goed om dat (ook) in `gateway_button` te doen?

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -357,11 +484,18 @@ function output_cell($rowspan, $data) {
 		</table>
-		<p><h3>Filter by dataset</h3></p>
+
+	<p><h3>Filter by dataset</h3></p>
```
Hier wordt de indentation stukgemaakt.

matthijskooijman commented on ttn_gw_data.json:
```diff
@@ -0,0 +1,263 @@
+{
+	"1dee0b64b020eec4": {
+       	 	"alias": "Meetjestad #1 (De WAR)",
```
Hier zit een indentation error.

matthijskooijman commented on ttn_gw_data.json:
```diff
@@ -0,0 +1,263 @@
+		"longitude": 5.39317479,
+		"altitude": 20
+	},
+	"1def5a0000000202": {
+		"alias": "Meetjestad #5 (Schothorst)",
+		"latitude": 52.17200,
+		"longitude": 5.39119,
+		"altitude": 5
+	},
+	"mjs-gateway-5": {
+		"alias": "Meetjestad #5 (Schothorst-W)",
+		"latitude": 52.17200,
+		"longitude": 5.39119,
+		"altitude": 5
+	},
```
Deze twee zou ik eruit halen, dit is een tijdelijke gateway voor ontwikkeling die overal kan zijn.

Peter-dM commented:
Hallo Matthijs,

Bedankt.

Mijn reacties staan er tussen.
Ik zal morgen de code aanpassen en de fork verversen.

/M. v.r gr.,/
*Peter*


On 28/04/2021 15:41, Matthijs Kooijman wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622157228>:
>
> >  		else
>   			$gw = $gw_id;
>   		$gw_html = htmlspecialchars($gw);
>   
> -		$filter_url_html = htmlspecialchars('?gateways=' . urlencode($gw_id) . '&show_other_gateways=1');
> +		#$filter_url_html = htmlspecialchars('?gateways=' . urlencode($gw_id) . '&show_other_gateways=1');
> +		$filter_url_html = htmlspecialchars('?gateways=' . urlencode($gw_id) . '&show_other_gateways=0');
>
> Heb je show_other_gateways hier bewust op 0 gezet, of was dat alleen 
> om te testen?
>
Ja, dat was bewust, volgens mij stond dit fout: ik zou denken dat hier 
alleen de berichten van deze TTN node van het laatste kwartier getoond 
horen te worden.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622159685>:
>
> > +				// fix old v2 record-layout:
> +				if (!array_key_exists('gateways', $metadata)) {
> +					$gateways = [];
> +	
> +					// Convert TTN staging to production format
> +					foreach ($metadata as $meta) {
> +						$meta['gtw_id'] = 'eui-' . strtolower($meta['gateway_eui']);
> +						$meta['snr'] = $meta['lsnr'];
> +						$gateways[] = $meta;
> +					}
> +	
> +					$metadata = [
> +						'gateways' => $gateways,
> +						'frequency' => $metadata[0]['frequency'],
> +						'data_rate' => $metadata[0]['datarate'],
> +						'coding_rate' => $metadata[0]['codingrate'],
> +					];
>
> Dit blokje code converteert van v1 naar v2, maar staat nu alleen 
> binnen het source == ttn.v2 blok, dus dat klopt nog niet. Volgens mij 
> moet het iets worden als:
>
> |if (ttn.v1) { // Bovenstaand blok } else if (ttn.v2) { $metadata = 
> $message['metadata']; } else if (ttn.v3) { // Jouw nieuwe code }|
|Inderdaad, ik heb over het hoofd gezien dat source ook 'ttn.v1' kan 
zijn.|||
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622161087>:
>
> >  				}
> -
> -				$metadata = [
> -					'gateways' => $gateways,
> -					'frequency' => $metadata[0]['frequency'],
> -					'data_rate' => $metadata[0]['datarate'],
> -					'coding_rate' => $metadata[0]['codingrate'],
> -				];
> -			}
> -
> +			} else if ($row['source'] == 'ttn.v3') {
> +				$metadata = $message['uplink_message']['rx_metadata'];
> +				$settings = $message['uplink_message']['settings'];
> +	
> +				if (!array_key_exists('gateway_ids', $metadata)) {
>
> Volgens mij is deze if bij ttn.v3 niet meer relevant, toch? Ik denk 
> dat die key nooit bestaat (dus if altijd true), dus die kan gewoon weg?
>
Die array_key bestaat volgens mij altijd, maarr de code wordt volgens 
mij altijd doorlopen (???).

Ik heb de regel weggehaald en ik zie geen verschil in het resultaat. 
Dus: kan inderdaad weg.

> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622161694>:
>
> > -				$metadata = [
> -					'gateways' => $gateways,
> -					'frequency' => $metadata[0]['frequency'],
> -					'data_rate' => $metadata[0]['datarate'],
> -					'coding_rate' => $metadata[0]['codingrate'],
> -				];
> -			}
> -
> +			} else if ($row['source'] == 'ttn.v3') {
> +				$metadata = $message['uplink_message']['rx_metadata'];
> +				$settings = $message['uplink_message']['settings'];
> +	
> +				if (!array_key_exists('gateway_ids', $metadata)) {
> +					$gateways = [];
> +	
> +					// Convert TTN staging to production format
>
> Dit commentaar is niet correct, staging was ttn.v1, production was 
> ttn.v2, dit converteert ttn.v3 naar ttn.v2 formaat voor verder gebruik 
> in de rest van het script.
>
Inderdaad.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622163125>:
>
> > -					'frequency' => $metadata[0]['frequency'],
> -					'data_rate' => $metadata[0]['datarate'],
> -					'coding_rate' => $metadata[0]['codingrate'],
> -				];
> -			}
> -
> +			} else if ($row['source'] == 'ttn.v3') {
> +				$metadata = $message['uplink_message']['rx_metadata'];
> +				$settings = $message['uplink_message']['settings'];
> +	
> +				if (!array_key_exists('gateway_ids', $metadata)) {
> +					$gateways = [];
> +	
> +					// Convert TTN staging to production format
> +					foreach ($metadata as $meta) {
> +						$meta['gtw_id'] = strtolower($meta['gateway_ids']['gateway_id']);
>
> Is deze strtolower nodig? Ik zou denk ik liever het id onveranderd 
> laten. In de ttn.v1 -> ttn.v2 conversie gebeurt het omdat de conventie 
> van v2 is EUI-based gateways een id van een lowercase eui te geven, 
> maar dat is hier niet van toepassing. Ik zou deze |strtolower| dus 
> weglaten.
>
Is ook mijn voorkeur. Ik gebruik het liefst bovenkast voor hex getallen. 
Dan moet de JSON ook aangepast worden denk ik.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622163601>:
>
> > +			} else if ($row['source'] == 'ttn.v3') {
> +				$metadata = $message['uplink_message']['rx_metadata'];
> +				$settings = $message['uplink_message']['settings'];
> +	
> +				if (!array_key_exists('gateway_ids', $metadata)) {
> +					$gateways = [];
> +	
> +					// Convert TTN staging to production format
> +					foreach ($metadata as $meta) {
> +						$meta['gtw_id'] = strtolower($meta['gateway_ids']['gateway_id']);
> +						$meta['gtw_eui'] = "";
> +						if (isset($meta['gateway_ids']['eui']))
> +							$meta['gtw_eui'] = strtolower($meta['gateway_ids']['eui']);
> +						if ($meta['gtw_id'] == 'packetbroker')
> +							$meta['gtw_id'] = 'unknown';
> +						if (! isset($meta['snr']))
>
> Hier staat een spatie achter de ! teveel.
>
OK
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622165375>:
>
> > +			if ($deb > 1) {
> +				print("310 ");
> +				print_r($gwdata);
> +				print("<br>");
> +			}
>
> Dit is debug code, die mag er uit voor we dit mergen.
>
OK
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622166926>:
>
> > +			if (! isset($gwdata['alias']))
> +				$gtw_id = $gwdata['gtw_id'];
>
> Kan |$gwdata['alias']| hier uberhaupt geset zijn? De code hierboven 
> zet nergens een alias, toch? Dus volgens mij kan de if weg, en de 
> |$gtw_id = ...| regel blijven.
>
Inderdaad.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622168706>:
>
> > +				if (isset($ttn_gw_data[$gtw_id]['alias'])) 
> +					$gwdata['alias'] = $ttn_gw_data[$gtw_id]['alias'];
>
> Deze regels hebben de verkeerde indentation (of er mist een |{| en |}| 
> als deze if binnen de if erboven hoort.
>
> Daarnaast: zijn deze twee regels uberhaupt nodig? Volgens mij staan ze 
> hieronder (regel 293/294) nog een keer identiek, en zijn ze (binnen de 
> |if(!unknown)| beter op hun plek.
>
Inderdaad, deze 2 regels kunnen weg.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622170015>:
>
> > +					if (isset($ttn_gw_data[$gtw_id]['alias'])) 
> +						$gwdata['alias'] = $ttn_gw_data[$gtw_id]['alias'];
>
> Ik zou deze twee regels een |if| naar buiten verplaatsen, zodat ze ook 
> gebeuren als er wel een longitude en latitude zijn.
>
Mee eens.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622170860>:
>
> > +		$gateways2 = array();
> +		foreach ($gateways as $gwdata) {
> +			if ($deb > 1) {
> +				print("310 ");
> +				print_r($gwdata);
> +				print("<br>");
> +			}
> +			// add alias if missing
> +			if (! isset($gwdata['alias']))
> +				$gtw_id = $gwdata['gtw_id'];
> +				if (isset($ttn_gw_data[$gtw_id]['alias']))
> +					$gwdata['alias'] = $ttn_gw_data[$gtw_id]['alias'];
> +			// add geo if missing
> +			if ($gwdata['gtw_id'] != 'unknown') {
> +				if (! isset($gwdata['longitude']) || ! isset($gwdata['latitude'])) {
> +					$gtw_id = $gwdata['gtw_id'];
>
> Dit gebeurt ook al hierboven (regel 286), dus volgens kan deze regel weg.
>
Inderdaad.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622171502>:
>
> > +			if ($deb > 0 && $gwdata['latitude'] != 0) {
> +				print("350 ");
> +				print_r($gwdata);
> +				print("<br>");
> +			}
>
> Ook weer debug code.
>
OK
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622172761>:
>
> > @@ -222,6 +273,51 @@ function output_cell($rowspan, $data) {
>   			$rowspan = 1;
>   		}
>   
> +		// enhance gwdata with alias and geo
> +		$gateways2 = array();
> +		foreach ($gateways as $gwdata) {
>
> Je zou hier een foreach by reference kunnen gebruiken, dan kun je 
> direct de originele |$gateways| aanpassen door dat |$gwdata| dan een 
> referentie wordt naar de data in |$gateways|, je hebt |$gateways2| dan 
> niet nodig. Volgens mij is de syntax daarvoor:
>
> foreach  ($gateways  as  &$gwdata) {
Inderdaad, dat is mooier, ik zal kijken waar de & moet.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622178594>:
>
> >  				if (array_key_exists('counter', $message)) {
>   					output_cell($rowspan, $message['counter']);
> +				} else if (array_key_exists('f_cnt', $message['uplink_message'])) {
> +					output_cell($rowspan, $message['uplink_message']['f_cnt']);
>
> Hier ben je eigenlijk afhankelijk van ttn.v1/v2 vs ttn.v3 een andere 
> element uit |$message| aan het pakken voor de frame counter. Het zou 
> denk ik netter zijn om alle ttn.v1/v2/v3 code naar bovenaan te 
> verplaatsen en daar |$metadata['counter']| oid te vullen, zodat je 
> hier gewoon direct alleen die waarde hoeft te gebruiken. Volgens mij 
> betekent dat ook dat je alleen in de v1/v2/v3 code bovenaan iets met 
> |$message| hoeft te doen, en je daarna alleen nog maar naar |$row|, 
> |$metadata| en |$gateways| hoeft te kijken.
>
Mee eens, eerder oplijnen is beter.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622180794>:
>
> > +				} else if (($gwdata['longitude'] == 0) && ($gwdata['latitude'] == 0)) {
> +					echo("  <td>" . $gw_id . "</td>\n");
>
> Je roept hier nu |gateway_button| niet aan als er geen locatie is voor 
> de gateway, maar dat betekent dat je dan ook niet op gateway id kunt 
> filteren. Daarnaast checkt |gateway_button| zelf al of er een locatie 
> is of niet, dus volgens mij kunnen deze twee regels weg.
>
OK
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622183842>:
>
> >  				echo("  <td>" . htmlspecialchars($gwdata["rssi"]) . "</td>\n");
> -				echo("  <td>" . htmlspecialchars($gwdata["snr"]) . "</td>\n");
> -				echo("  <td>" . htmlspecialchars($metadata["frequency"]) . "Mhz, " . htmlspecialchars($metadata["data_rate"]) . ", " .htmlspecialchars($metadata["coding_rate"]) . "CR</td>\n");
> +
> +				if ($gwdata["snr"] == $null_snr)
> +					echo("  <td>x</td>\n");
> +				else
> +					echo("  <td>" . htmlspecialchars($gwdata["snr"]) . "</td>\n");
> +
> +				if ($row['source'] == 'ttn.v2')
> +					echo("  <td>" . htmlspecialchars($metadata["frequency"]) . "MHz, " . htmlspecialchars($metadata["data_rate"]) . ", " .htmlspecialchars($metadata["coding_rate"]) . "CR</td>\n");
> +				if ($row['source'] == 'ttn.v3')
> +					echo("  <td>" . htmlspecialchars($metadata["frequency"]) . "MHz, SF" . htmlspecialchars($metadata["spreading_factor"]) . "BW" . htmlspecialchars($metadata["data_rate"]) . ", " .htmlspecialchars($metadata["coding_rate"]) . "CR</td>\n");
>
> Ik zou deze conversie (|SF|, |spreading_factor|, |BW| en |data_rate| 
> aan elkaar plakken) bovenaan doen, waar je toch al v3 apart behandeld 
> en het resultaat in |$metadata["data_rate"]| stoppen. Je kunt dan hier 
> gewoon onafhankelijk van de source versie gewoon 
> |$metadata["data_rate"]| direct gebruiken (dus zoals het er origineel 
> stond).
>
OK
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622188059>:
>
> > @@ -357,11 +484,18 @@ function output_cell($rowspan, $data) {
>   					}
>   				}
>   				$stationcount = count($stations);
> -				echo("<tr><td>".gateway_button($gw_id)."</td><td>".htmlspecialchars($messagecount)."</td><td>".htmlspecialchars($stationcount)."</td><td>".$stationlist."</td></tr>\n");
> +
> +				$alias = "&nbsp;";
> +				if (isset($ttn_gw_data[$gw_id]['alias']))
> +					$alias = $ttn_gw_data[$gw_id]['alias'];
> +				$url = "https://ttnmapper.org/gateways/?gateway=$gw_id&startdate=&enddate=&gateways=on&points=on";
>
> Goed idee om te linken naar ttnmapper, misschien goed om dat (ook) in 
> |gateway_button| te doen?
>
OK, ik vind de geomap ook nuttig, misschien kunnen we allebei tonen.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622188537>:
>
> >  			}
>   		?>
>   		</table>
> -		<p><h3>Filter by dataset</h3></p>
> +
> +	<p><h3>Filter by dataset</h3></p>
>
> Hier wordt de indentation stukgemaakt.
>
OK, ik bedoelde dit als een paginadeling
> ------------------------------------------------------------------------
>
> In ttn_gw_data.json 
> <https://github.com/meetjestad/website/pull/15#discussion_r622188712>:
>
> > @@ -0,0 +1,263 @@
> +{
> +	"1dee0b64b020eec4": {
> +       	 	"alias": "Meetjestad #1 (De WAR)",
>
> Hier zit een indentation error.
>
OK
> ------------------------------------------------------------------------
>
> In ttn_gw_data.json 
> <https://github.com/meetjestad/website/pull/15#discussion_r622189571>:
>
> > +	"1def5a0000000202": {
> +		"alias": "Meetjestad #5 (Schothorst)",
> +		"latitude": 52.17200,
> +		"longitude": 5.39119,
> +		"altitude": 5
> +	},
> +	"mjs-gateway-5": {
> +		"alias": "Meetjestad #5 (Schothorst-W)",
> +		"latitude": 52.17200,
> +		"longitude": 5.39119,
> +		"altitude": 5
> +	},
>
> Deze twee zou ik eruit halen, dit is een tijdelijke gateway voor 
> ontwikkeling die overal kan zijn
>
Deze had ik nodig om v3 te testen. Ik zou ze willen laten staan tot er 
andere zijn om mee te testen.
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622176093>:
>
> > +				// fix old v2 record-layout:
> +				if (!array_key_exists('gateways', $metadata)) {
> +					$gateways = [];
> +	
> +					// Convert TTN staging to production format
> +					foreach ($metadata as $meta) {
> +						$meta['gtw_id'] = 'eui-' . strtolower($meta['gateway_eui']);
> +						$meta['snr'] = $meta['lsnr'];
> +						$gateways[] = $meta;
> +					}
> +	
> +					$metadata = [
> +						'gateways' => $gateways,
> +						'frequency' => $metadata[0]['frequency'],
> +						'data_rate' => $metadata[0]['datarate'],
> +						'coding_rate' => $metadata[0]['codingrate'],
> +					];
>
> Of misschien is het nog explicieter om in het geval van ttn.v2 het uit 
> te schrijven:
>
> |$metadata = [ 'gateways' => $message['metadata']['gateways'], 
> 'frequency' => $message['metadata']['frequency'], 'data_rate' => 
> $message['metadata']['datarate'], 'coding_rate' => 
> $message['metadata']['codingrate'], ];|
|OK, de reden waarom je zoiets doet is |||om de code compact te houden
> ------------------------------------------------------------------------
>
> In data/sensors_recent.php 
> <https://github.com/meetjestad/website/pull/15#discussion_r622165815>:
>
> > +			} else if ($row['source'] == 'ttn.v3') {
> +				$metadata = $message['uplink_message']['rx_metadata'];
> +				$settings = $message['uplink_message']['settings'];
> +	
> +				if (!array_key_exists('gateway_ids', $metadata)) {
> +					$gateways = [];
> +	
> +					// Convert TTN staging to production format
> +					foreach ($metadata as $meta) {
> +						$meta['gtw_id'] = strtolower($meta['gateway_ids']['gateway_id']);
> +						$meta['gtw_eui'] = "";
> +						if (isset($meta['gateway_ids']['eui']))
> +							$meta['gtw_eui'] = strtolower($meta['gateway_ids']['eui']);
> +						if ($meta['gtw_id'] == 'packetbroker')
> +							$meta['gtw_id'] = 'unknown';
> +						if (! isset($meta['snr']))
>
> Idem voor een aantal andere plekken elders in het script.
>
... volledig uitschrijven van het record, begrijp ik.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/meetjestad/website/pull/15#pullrequestreview-647034415>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AR5Z2DOOI2SDBXLUUZR6LJTTLAGA7ANCNFSM43TIJLWQ>.
>

matthijskooijman commented:
Ah, ik zie net pas dat je al weer even geleden een volgende versie gepushed had. Allereerst twee procesmatige suggesties:
 - Als je een nieuwe versie hebt gepushed die klaar is voor review, is het vaak goed om daar ook even expliciet een berichtje op github over achter te laten. Op zich krijg ik doorgaans wel een mailtje als er iets nieuws gepushed wordt, maar omdat het nog wel eens voorkomt dat mensen al tussendoor even wat pushen voor het helemaal klaar is, kijk ik meestal niet zo goed naar dat soort mailtjes.
 - Mijn reacties staan in de github webinterface netjes onderverdeeld in aparte commentaartjes, met de bijbehorende code er direct boven. Als je ook via de webinterface je reacties geeft (ipv via e-mail wat je nu denk ik gedaan hebt), is het net wat makkelijker om de boel uit elkaar te houden (de reacties over 1 onderwerp worden dan vanzelf in een draadje gegroepeerd, zie hierboven in de webinterface).


> >  Heb je show_other_gateways hier bewust op 0 gezet, of was dat alleen
 om te testen?

> Ja, dat was bewust, volgens mij stond dit fout: ik zou denken dat hier
alleen de berichten van deze TTN node van het laatste kwartier getoond
horen te worden.

Ik snap niet helemaal wat je hier bedoelt met "berichten van deze TTN node", aangezien deze linkjes filteren op gateway, niet op node? En de "show_other_gateways" heeft geen invloed op de filtering, als je 1 gateway selecteert zie je altijd alleen de berichten die door die gateway zijn opgevangen. Echter, als show_other_gatways 1 is, dan zie je bij die berichten ook de info van andere gateways die hetzelfde bericht *ook* ontvangen hebben, als ie 0 is zie je alleen info van de ene gateway waar je op filtert. In mijn beleving is het vaak juist wel handig om te zien welke andere gateways berichten nog meer hebben ontvangen, vandaar dat ie op 1 stond, maar wel expliciet in de url zodat mensen hem zien en evt op 0 kunnen zetten als ze dat willen.

> |OK, de reden waarom je zoiets doet is |||om de code compact te houden

Het voorstel wat ik daar deed (voor ttn.v2 het toewijzen van `$metadata` helemaal uitschrijven) maakt de code niet compacter, maar zorgt er voor dat de structuur van ttn.v1/v2/v3 vergelijkbaarder is (ze eindigen dan allemaal met het toewijzen van alle afzonderlijke elementen van `$metadata`).

>  Idem voor een aantal andere plekken elders in het script.
> > ... volledig uitschrijven van het record, begrijp ik.

Nee, dit ging over extra spaties achter `!`. Mogelijk dat dit in de e-mailnotificaties niet duidelijk was, maar als je dit commentaar hierboven in de webinterface opzoekt is het waarschijnlijk helderder.

*edit: Maar zo te zien had je dat wel begrepen, want al die spaties zijn weg :-D*

Tot zover even de reactie op jouw reacties, ik ga nu je nieuw commit bekijken. Ik zal in github de opmerkingen die zijn opgelost als "resolved" markeren, en waar nog wat nodig is op de bestaande comments reageren.

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -154,29 +144,28 @@ function node_button($id) {
-    <a class="dropdown-item gw_location" href="$location_url_html">Show gateway location</a>
+    <a class="dropdown-item gw_filter"   href="$filter_url_html">Show only messages received by $gw_html</a>
+    <a class="dropdown-item gw_location" href="$location_url_html">Show gateway on map</a>
+    <a class="dropdown-item gw_loc_ttn"  href="$location_ttn_html">Show gateway on TTN</a>
```
This should probably say "TTN Mapper" instead of "TTN"

Peter-dM replied:
Agreed, fixed

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -193,59 +182,64 @@ function output_cell($rowspan, $data) {
-			if ($row['source'] == 'ttn.v2') {
+			if ($row['source'] == 'ttn.v1') {
 				$metadata = $message['metadata'];
+				$metadata['counter'] = $message['counter'];
```
Now that `$metadata` is assigned explicitly a bit further down, these two lines are no longer needed (`$metadata` will be completely overridden).

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -193,59 +182,64 @@ function output_cell($rowspan, $data) {
+
+			} else if ($row['source'] == 'ttn.v2') {
+				$metadata = $message['metadata'];
+				$metadata['counter'] = $message['counter'];
```
Here I suggested to explicitly assign all elements of `$metada` as well (exactly like you now did for `ttn.v1`).

Peter-dM replied:
fixed as suggested

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -193,59 +182,64 @@ function output_cell($rowspan, $data) {
+				$metadata['counter'] = $message['counter'];
+
 			} else if ($row['source'] == 'ttn.v3') {
 				$metadata = $message['uplink_message']['rx_metadata'];
```
Deze regel kan ook weg, aangezien verderop `$metadata` helemaal overschreven wordt.

Peter-dM replied:
Werd wel gebruikt om $gateways te vullen, die worden nu gevuld uit $message

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -193,59 +182,64 @@ function output_cell($rowspan, $data) {
+			}	// if ttn.v1 else ttn.v2 else ttn.v3
 
 			$gateways = $metadata['gateways'];
+			if (count($gateways) > 3) $gateways = array($gateways[0], $gateways[1], $gateways[2]);	// now only showing 3 best-receiving gateways
```
I don't think this is a good idea, since this silently hides information. If we want to make the interface more compact (which would make sense, given there are so much gateways nowadays), I would rather suggest doing this on the browser by collapsing the extra gateways and allowing the user to expand the list. But I think this is out of scope for this PR, so please remove this again and then we can figure this out in a separate PR later.

Peter-dM replied:
I suggest a URL parameter like: ...&maxgw=3

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -300,23 +285,15 @@ function output_cell($rowspan, $data) {
 		}
-		
-		$gateways = $gateways2;
+		unset($gwdata);
```
Waarom deze unset? Om te voorkomen dat die variabele per ongeluk nog gebruikt wordt?

Peter-dM replied:
Inderdaad. Hij is nodig om de loop goed te laten werken.
Ik beschouw deze unset als onderdeel van de PHP syntax

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -371,21 +348,20 @@ function output_cell($rowspan, $data) {
 				if ($row['firmware_version'] === null)
 					output_cell($rowspan, '< v1');
-				else if ($row['firmware_version'] !== 255)
+				else if ($row['firmware_version'] !== 255 && $row['firmware_version'] !== '255')
```
Hm, blijkbaar komen er strings uit de database en geen integers en heb ik dit nooit getest :-)

Ik zou het dan mooier vinden omde integerversie helemaal weg te laten, en gewoon alleen tegen `'255'` te testen (of anders met `!=` testen om onafhankelijk van type te testen, maar dat heeft niet mijn voorkeur).

Peter-dM replied:
aangepast, er staat nu alleen nog '255'

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -138,63 +142,104 @@ function node_button($id) {
 		$gw_html = htmlspecialchars($gw);
 
-		$filter_url_html = htmlspecialchars('?gateways=' . urlencode($gw_id) . '&show_other_gateways=1');
+		$filter_url_html = htmlspecialchars('?gateways=' . urlencode($gw_id) . '&show_other_gateways=0');	// now show_other_gateways=0
```
As discussed, I'd rather keep this at show_other_gateways=1. Also, you added a comment that is really about a *change* you made, rather than the current code (or in a way it is about the current code, but it is entirely obvious and just repeats what the code alrady says), which often is not that useful. Such comments are better put into commit messages, but in this case it can just be omitted IMHO.

Peter-dM replied:
OK, rolled back

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -371,21 +348,20 @@ function output_cell($rowspan, $data) {
-					output_cell($rowspan, $message['counter']);
-				} else if (array_key_exists('f_cnt', $message['uplink_message'])) {
-					output_cell($rowspan, $message['uplink_message']['f_cnt']);
+				if (array_key_exists('counter', $metadata)) {
```
Volgens mij kan deze if weg, aangezien `$metadata['counter']` bovenaan nu altijd een waarde krijgt. Ik heb ook nog even in de database gechecked, maar zo te zien hebben *alle* ttn.v1 en ttn.v2 berichten ook een counter, dus ik weet niet zo goed waarom deze if er in de eerste plaats was. Dus kan nu zeker weg (en de `-` in de else dus ook).

Peter-dM replied:
Aangepast.

matthijskooijman reviewed (changes requested):
Ik heb nog een keer over de code heen gekeken, en complimenten voor de nauwkeurigheid waarmee je mijn stortvloed aan commentaar hebt verwerkt, niets gemist :-)

Ik heb nog wel een paar nieuwe kleinere aanmerkingen geplaatst, wil je daar nog even naar kijken? Alles van de vorige review heb ik als resolved gemarkeerd (en is op github nu dus ingeklapt), voor al mijn aanmerkingen heb ik nu een nieuw commenteertje toegevoegd. Volgens mij zijn we er bijna :-)

Ik zie trouwens dat ik uit automatisme weer in het Engels ben gaan reageren, kijk maar even of je Engels of Nederlands fijner vindt om weer te reageren :-)

matthijskooijman commented:
> Als je een nieuwe versie hebt gepushed die klaar is voor review, is het vaak goed om daar ook even expliciet een berichtje op github over achter te laten.

Ik zie net dat je me via Element al een prima berichtje had gestuurd, dus daarna ben ik het gewoon zelf vergeten :-p Voordeel van een berichtje op github is dat ie mijn inbox terecht komt en ik hem kan laten staan tot ik er naar gekeken heb, dat wel :-)

matthijskooijman commented on data/sensors_recent.php:
```diff
@@ -138,63 +142,104 @@ function node_button($id) {
+					'frequency'   => $message['uplink_message']['settings']['frequency'] / 1000000.,
+					'coding_rate' => $message['uplink_message']['settings']['coding_rate'],
+					'data_rate'   => $data_rate,
+					'counter'     => $message['uplink_message']['f_cnt'],
```
Turns out that in v3, a `f_cnt` of 0 is currently being emitted from the generated JSON (see https://github.com/TheThingsNetwork/lorawan-stack/issues/3874), so we should handle that here. I would suggest making this change:
```suggestion
					'counter'     => isset($message['uplink_message']['f_cnt']) ? $message['uplink_message']['f_cnt'] : 0,
```

matthijskooijman replied:
This can be seen in the messags from 2022 in the past evening, I've been resetting it a lot:

![image](https://user-images.githubusercontent.com/194491/119060435-7315a080-b9d2-11eb-8ae7-8327c24d68f2.png)

Peter-dM committed:
sensors_recent: Remove redundant if

It turns out the counter is always present in messages, so no need to
check here.

Peter-dM committed:
sensors_recent: Add TTN Mapper link for gateways

Peter-dM committed:
sensors_recent: Add additional gateway info from external JSON file

Previously, a more readable description for some gateways was stored in
the code directly. This is now moved into an external file, and also
gateway locations are added to work around locations being missing in
the TTNv3 MQTT stream, see
https://github.com/TheThingsNetwork/lorawan-stack/issues/4218

Peter-dM committed:
sensors_recent: Fix wrong casing in DateTimeZone

This probably did not cause any problems, since PHP is probably case
insensitive here.

Peter-dM committed:
sensors_recent: Fix development firmware version check

The database produced strings, but this checked against an integer.

Peter-dM committed:
sensors_recent: Sort "Messages per node" list descending

Peter-dM committed:
sensors_recent: Support TTNv3

Peter-dM committed:
sensors_recent: Fix casing of MHz

Peter-dM committed:
sensors_recent: Show source / TTN version for messages

matthijskooijman committed:
sensors_recent: Allow filtering based on source

matthijskooijman commented:
Code looks good now, thanks! I've taken your changes and split them into different logical commits for easier future reference. I've force pushed this to your master branch to update this PR, so pull carefully.

I've left out a few of your changes, in particular:
 - I discovered that a missing SNR actually means the SNR is exactly 0 (similar to how f_cnt is omitted when it is 0), so I adapted to code to just use 0 and remove the null_snr handling.
 - I moved the "unknown" handling into gateway_button, to simplify things a bit.
 - There was some unused $deb variable and some unused $metadata assignment
 - I left out some whitespace-only changes and removed some trailing whitespace
 - I left out the extra gw alias column in the gateway table at the bottom and the ttn mapper link, since gateway_button already displays those.

See below for a full diff of the changes compared to your version of the code.

I also added a source= filter (e.g. `&source=ttn.v2` in the URL) as we discussed on Element (I needed it for some quick testing, so I just added it right away).

With that, I'm going to merge and deploy this. Thanks for all your work so far!

matthijskooijman commented:
Here's the diff of changes:

```diff

diff --git a/data/sensors_recent.php b/data/sensors_recent.php
index 059fa82..5e047f1 100644
--- a/data/sensors_recent.php
+++ b/data/sensors_recent.php
@@ -24,11 +24,10 @@ function haversineGreatCircleDistance(
 	  $lonDelta = $lonTo - $lonFrom;
 
 	  $angle = 2 * asin(sqrt(pow(sin($latDelta / 2), 2) +
-		cos($latFrom) * cos($latTo) * pow(sin($lonDelta / 2), 2)));
+	    cos($latFrom) * cos($latTo) * pow(sin($lonDelta / 2), 2)));
 	  return $angle * $earthRadius;
 	}
 
-
 	header('Content-Type: text/html; charset=utf-8');
 
 	if (file_exists('../sensorsets.json')) $sensorsets = json_decode(file_get_contents('../sensorsets.json'), true);
@@ -74,9 +73,6 @@ function haversineGreatCircleDistance(
 				<th>Radiosettings</th>
 			</tr>
 <?php
-	$null_snr = -99;
-	$deb = 0;
-
 	// connect to database
 	include ("../connect.php");
 	$database = Connection();
@@ -90,13 +86,9 @@ function haversineGreatCircleDistance(
 	// Compatibility with existing links
 	if (isset($_GET['sensor']))
 		$_GET['sensors'] = $_GET['sensor'];
-
 	if (isset($_GET['gateway']))
 		$_GET['gateways'] = $_GET['gateway'];
 
-	if (isset($_GET['deb']))
-		$deb = $_GET['deb'];
-
 	if (isset($_GET['sensors'])) {
 		$sensors = explode(',', $_GET['sensors']);
 		foreach ($sensors as &$id)
@@ -107,7 +99,6 @@ function haversineGreatCircleDistance(
 		$WHERE = "";
 	}
 
-
 	$gateways_filter = false;
 	if (isset($_GET['gateways']))
 		$gateways_filter = explode(',', $_GET['gateways']);
@@ -125,7 +116,6 @@ function haversineGreatCircleDistance(
 	$distances_per_gateway = array();
 	$messagecount = 0;
 
-
 	function node_button($id) {
 		$id = intval($id);
 		return <<<EOF
@@ -142,8 +132,10 @@ function node_button($id) {
 EOF;
 	}
 
-
 	function gateway_button($gw_id, $gw_data = null) {
+
+		if ($gw_id == 'unknown')
+			return '<i>unknown</i>';
 		if (isset($gw_data['alias']))
 			$gw = $gw_data['alias'];
 		else
@@ -171,23 +163,17 @@ function gateway_button($gw_id, $gw_data = null) {
 EOF;
 	}
 
-
 	function output_cell($rowspan, $data) {
 		echo("  <td rowspan=\"$rowspan\"> " . $data . "</td>\n");
 	}
 
-
-	// main loop:
 	while($row = $result->fetch_array(MYSQLI_ASSOC)) {
 		if ($row['message']) {
 			$message = json_decode($row['message'], true);
 
 			if ($row['source'] == 'ttn.v1') {
-				$metadata = $message['metadata'];
-				$metadata['counter'] = $message['counter'];
-
 				$gateways = [];
-	
+
 				// Convert TTN v1=staging to v2=production format
 				foreach ($metadata as $meta) {
 					$meta['gtw_id'] = 'eui-' . strtolower($meta['gateway_eui']);
@@ -195,7 +181,6 @@ function output_cell($rowspan, $data) {
 					$gateways[] = $meta;
 				}
 
-
 				$metadata = [
 					'gateways'    => $message['metadata']['gateways'],
 					'frequency'   => $message['metadata']['frequency'],
@@ -214,7 +199,7 @@ function output_cell($rowspan, $data) {
 				];
 
 			} else if ($row['source'] == 'ttn.v3') {
-	
+
 				// Convert TTN v3 to v2 format
 
 				// Retrieve gateway details
@@ -224,14 +209,17 @@ function output_cell($rowspan, $data) {
 					$meta['gtw_eui'] = "";
 					if (isset($meta['gateway_ids']['eui']))
 						$meta['gtw_eui'] = $meta['gateway_ids']['eui'];
+					// Gateways bridge from ttnv2 do not include their id, see:
+					// https://github.com/packetbroker/api/issues/23#issuecomment-852021215
 					if ($meta['gtw_id'] == 'packetbroker')
 						$meta['gtw_id'] = 'unknown';
-					if (!isset($meta['snr'])) 
-						$meta['snr'] = $null_snr;
+					// Default values (zeroes) are omitted: https://github.com/TheThingsNetwork/lorawan-stack/issues/3874
+					if (!isset($meta['snr']))
+						$meta['snr'] = 0;
 
 					$gateways[] = $meta;
 				}
-	
+
 				$spreading_factor = $message['uplink_message']['settings']['data_rate']['lora']['spreading_factor'];
 				$bandwidth =        $message['uplink_message']['settings']['data_rate']['lora']['bandwidth'] / 1000;
 				$data_rate = sprintf("SF%sBW%s", $spreading_factor, $bandwidth);
@@ -240,9 +228,10 @@ function output_cell($rowspan, $data) {
 					'frequency'   => $message['uplink_message']['settings']['frequency'] / 1000000.,
 					'coding_rate' => $message['uplink_message']['settings']['coding_rate'],
 					'data_rate'   => $data_rate,
+					// Default values (zeroes) are omitted: https://github.com/TheThingsNetwork/lorawan-stack/issues/3874
 					'counter'     => isset($message['uplink_message']['f_cnt']) ? $message['uplink_message']['f_cnt'] : 0,
 				];
-			}	// if ttn.v1 else ttn.v2 else ttn.v3
+			}
 
 			$gateways = $metadata['gateways'];
 			if ($gateways_filter) {
@@ -277,15 +266,15 @@ function output_cell($rowspan, $data) {
 			$gtw_id = $gwdata['gtw_id'];
 			if ($gwdata['gtw_id'] != 'unknown') {
 				// add alias
-				if (isset($ttn_gw_data[$gtw_id]['alias'])) 
+				if (isset($ttn_gw_data[$gtw_id]['alias']))
 					$gwdata['alias'] = $ttn_gw_data[$gtw_id]['alias'];
 				// add geo if missing
 				if (!isset($gwdata['longitude']) || !isset($gwdata['latitude'])) {
-					if (isset($ttn_gw_data[$gtw_id]['latitude'])) 
+					if (isset($ttn_gw_data[$gtw_id]['latitude']))
 						$gwdata['latitude'] = $ttn_gw_data[$gtw_id]['latitude'];
-					if (isset($ttn_gw_data[$gtw_id]['longitude'])) 
+					if (isset($ttn_gw_data[$gtw_id]['longitude']))
 						$gwdata['longitude'] = $ttn_gw_data[$gtw_id]['longitude'];
-					if (isset($ttn_gw_data[$gtw_id]['altitude'])) 
+					if (isset($ttn_gw_data[$gtw_id]['altitude']))
 						$gwdata['altitude'] = $ttn_gw_data[$gtw_id]['altitude'];
 				}
 			}
@@ -319,20 +308,15 @@ function output_cell($rowspan, $data) {
 				$datetime = DateTime::createFromFormat('Y-m-d H:i:s', $row['timestamp'], new DateTimeZone('UTC'));
 				$datetime->setTimeZone(new DateTimeZone('Europe/Amsterdam'));
 				output_cell($rowspan, $datetime->format('Y-m-d H:i:s'));
-
 				output_cell($rowspan, $row["temperature"] . "°C");
-
 				output_cell($rowspan, $row["humidity"] . "%");
-
 				output_cell($rowspan, $row["lux"] . ($row["lux"]>0?" lux":""));
-
 				if ($row["pm2_5"] == 0xffff)
 					output_cell($rowspan, "read error");
 				else if ($row["pm2_5"])
 					output_cell($rowspan, $row["pm2_5"] . " μg/m³");
 				else
 					output_cell($rowspan, "");
-
 				if ($row["pm10"] == 0xffff)
 					output_cell($rowspan, "read error");
 				else if ($row["pm10"])
@@ -379,26 +363,14 @@ function output_cell($rowspan, $data) {
 				if ($row['latitude'] && array_key_exists('latitude', $gwdata) && $gwdata['latitude']) {
 					$distance = haversineGreatCircleDistance($row['latitude'], $row['longitude'], $gwdata['latitude'], $gwdata['longitude']);
 				}
-
 				$gw_id = $gwdata['gtw_id'];
-				if ($gw_id == "unknown") {
-					echo("  <td><I>" . $gw_id . "</i></td>\n");
-				} else {
-					echo("  <td>" . gateway_button($gw_id, $gwdata) . "</td>\n");
-				}
-
+				echo("  <td>" . gateway_button($gw_id, $gwdata) . "</td>\n");
 				if ($distance)
-					echo("  <td>" . round($distance / 1000, 3) . " km</td>\n");
+					echo("  <td>" . round($distance / 1000, 3) . "km</td>\n");
 				else
 					echo("  <td>-</td>\n");
-
 				echo("  <td>" . htmlspecialchars($gwdata["rssi"]) . "</td>\n");
-
-				if ($gwdata["snr"] == $null_snr)
-					echo("  <td>x</td>\n");
-				else 
-					echo("  <td>" . htmlspecialchars($gwdata["snr"]) . "</td>\n");
-
+				echo("  <td>" . htmlspecialchars($gwdata["snr"]) . "</td>\n");
 				echo("  <td>" . htmlspecialchars($metadata["frequency"]) . "MHz, " . htmlspecialchars($metadata["data_rate"]) . ", " .htmlspecialchars($metadata["coding_rate"]) . "CR</td>\n");
 
 				if (!array_key_exists($gw_id, $stations_per_gateway)) {
@@ -418,8 +390,6 @@ function output_cell($rowspan, $data) {
 	?>
 		</table>
 		<p>Message count: <?= $messagecount ?><br>Node count: <?= count($count_per_station)?></p>
-
-
 		<p><h3>Messages per node</h3></p>
 		<table class="table table-bordered table-striped table-sm" style="width: calc(100% - 20px); margin: 10px">
 		<tr class="bg-info"><th>Number of messages in list above</th><th>Nodes</th></tr>
@@ -431,7 +401,7 @@ function output_cell($rowspan, $data) {
 
 				$stations_per_count[$count][] = $station;
 			}
-			krsort($stations_per_count);	// now sorting descending
+			krsort($stations_per_count);
 			foreach($stations_per_count as $count => $stations) {
 				$stationlist = implode("", array_map('node_button', $stations));
 				echo("<tr><td>".htmlspecialchars($count)."</td><td>".$stationlist."</td></tr>\n");
@@ -439,10 +409,9 @@ function output_cell($rowspan, $data) {
 		?>
 		</table>
 
-
-		<p><h3>Statistics per gateway</h3></p>
+	<p><h3>Statistics per gateway</h3></p>
 		<table class="table table-bordered table-striped table-sm" style="width: calc(100% - 20px); margin: 10px">
-		<tr class="bg-info"><th>Gateway</th><th>Alias</th><th>Number of messages</th><th>Number of nodes</th><th>Nodes</th></tr>
+		<tr class="bg-info"><th>Gateway</th><th>Number of messages</th><th>Number of nodes</th><th>Nodes</th></tr>
 		<?php
 			arsort($messagecount_per_gateway);
 			foreach($messagecount_per_gateway as $gw_id => $messagecount) {
@@ -456,18 +425,10 @@ function output_cell($rowspan, $data) {
 					}
 				}
 				$stationcount = count($stations);
-
-				$alias = "&nbsp;";
-				if (isset($ttn_gw_data[$gw_id]['alias']))
-					$alias = $ttn_gw_data[$gw_id]['alias'];
-				$ttn = "https://ttnmapper.org/gateways/?gateway=$gw_id&startdate=&enddate=&gateways=on&points=on";
-				if ($gw_id != 'unknown') 
-					echo("<tr><td><A HREF=$ttn>" . $gw_id . "</a></td><td>" . $alias . "</td><td>".htmlspecialchars($messagecount)."</td><td>".htmlspecialchars($stationcount)."</td><td>".$stationlist."</td></tr>\n");
+				echo("<tr><td>".gateway_button($gw_id)."</td><td>".htmlspecialchars($messagecount)."</td><td>".htmlspecialchars($stationcount)."</td><td>".$stationlist."</td></tr>\n");
 			}
 		?>
 		</table>
-
-
 		<p><h3>Filter by dataset</h3></p>
 		<ul>
 		<?php foreach ($sensorsets as $id => $set) { ?>
@@ -476,4 +437,3 @@ function output_cell($rowspan, $data) {
 		</ul>
 	</body>
 </html>
-
```
